### PR TITLE
py2->py3: nbd_client_manager.py

### PR DESCRIPTION
**Pull request by stephenchengCloud**
- Create new directory `python3/libexec` for `nbd_client_manager.py`
- Adjust Makefile and CI check
- Port  `nbd_client_manager.py` to python3
- Added unit test
- Manually tested
  - connect an nbd server: 
image
  - disconnect the nbd server, removed the connect info: 
image
- XenRT test
  - 3944595: Debian passed
  - 3944596: corosync4hostsHAP1 passed
  - 3944597: bvt-pool.seq passed
 
xensource.log from 3944597:
>  Mar  5 07:29:02 conapuk-25-01c xapi: [debug||6394 HTTP 127.0.0.1->:::80|pool.enable_ha R:ae4b797f50b7|helpers] about to call script without a timeout: /opt/xensource/bin/static-vdis attach b8d9978f-62a9-4919-ae6b-dbef1c5962e4
> Mar  5 07:29:02 conapuk-25-01c xapi: [debug||6394 HTTP 127.0.0.1->:::80|pool.enable_ha R:ae4b797f50b7|helpers] /opt/xensource/bin/static-vdis attach b8d9978f-62a9-4919-ae6b-dbef1c5962e4 succeeded [ output = '/etc/xensource/static-vdis/1/disk\x0A' ]
> ...
> Mar  5 07:37:20 conapuk-25-01c xapi: [ info||7003 /var/lib/xcp/xapi|host.ha_release_resources R:5ea616b86a5f|static_vdis] permanent_vdi_detach: vdi-uuid = 3b9846a7-63b2-48e4-b021-efc9528464f8
> Mar  5 07:37:20 conapuk-25-01c xapi: [debug||7003 /var/lib/xcp/xapi|host.ha_release_resources R:5ea616b86a5f|helpers] about to call script without a timeout: /opt/xensource/bin/static-vdis detach 3b9846a7-63b2-48e4-b021-efc9528464f8
> Mar  5 07:37:20 conapuk-25-01c xapi: [debug||7003 /var/lib/xcp/xapi|host.ha_release_resources R:5ea616b86a5f|helpers] /opt/xensource/bin/static-vdis detach 3b9846a7-63b2-48e4-b021-efc9528464f8 succeeded [ output = '' ]
> 
call chain: 
https://github.com/xapi-project/xen-api/blob/master/scripts/static-vdis#L279
attach()->connect_nbd()->nbd_client_manager.connect_nbd()
https://github.com/xapi-project/xen-api/blob/master/scripts/static-vdis#L328
detach()->disconnect_nbd_device()->nbd_client_manager.disconnect_nbd_device()

### CODE REVIEW

**Comment by stephenchengCloud** [comment]
Pylint checks "using f-string", but our build system uses python3.6 which doesn't support f-string.
So disable the check.

**Comment by stephenchengCloud** [comment]
Re-built and re-tested based on the latest commit:
3946459: bvt-pool
3946460: Debian

**Comment by psafont** [comment]
Looks like 3.6 needs to be tested against since XS8 uses it. Can this version be enabled in CI, here? https://github.com/xapi-project/xen-api/blob/master/.github/workflows/main.yml#L21

**Comment by stephenchengCloud** [comment]
> Looks like 3.6 needs to be tested against since XS8 uses it. Can this version be enabled in CI, here? https://github.com/xapi-project/xen-api/blob/master/.github/workflows/main.yml#L21

I'll try it.

**Comment by stephenchengCloud** [comment]
Ocaml test failed due to "hex not found" 
Need to sync with master branch first.

**Comment by stephenchengCloud** [comment]
Rebase after syncing with the master branch.
Since no code changed. Just manually tested it:
image

File: pyproject.toml [hunk 1]
```diff
@@ -34,6 +34,13 @@ profile = "black"
 combine_as_imports = true
 ensure_newline_before_comments = false
 
+[tool.pylint.messages_control]
+disable = [
+    "missing-function-docstring",
```
Comment (liulinC) [inline pyproject.toml:39]: The original code has no doc string at all, 
It make sense to me to disable it ATM.

File: python3/libexec/nbd_client_manager.py [hunk 2]
```diff
@@ -1,52 +1,66 @@
...
+
     def __init__(self, path):
         self._path = path
         self._lock_file = None
 
     def _lock(self):
         """Acquire the lock"""
         flags = fcntl.LOCK_EX
-        self._lock_file = open(self._path, 'w+')
+        # pylint: disable=consider-using-with
```
Comment (liulinC) [inline python3/libexec/nbd_client_manager.py:62]: can not use `with` state here as it is a file lock, need to close the file at proper place explicitly.
So it make sense to me to disable this check.

**Review by liulinC — APPROVED** [review]

**Comment by psafont** [comment]
This needs a rebase with all conflicts solved

File: python3/libexec/nbd_client_manager.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ class NbdDeviceNotFound(Exception):
 
     def __init__(self, nbd_device):
         super().__init__(
-            f"NBD device '{nbd_device}' does not exist"
+            "NBD device '{}' does not exist".format(nbd_device)
```
Comment (psafont) [inline python3/libexec/nbd_client_manager.py:48]: f-strings were introduced in 3.6, am I missing something? https://docs.python.org/3.6/whatsnew/3.6.html#pep-498-formatted-string-literals

**Review by psafont — APPROVED** [review]
Testing looks really good, thanks for the effort!

File: python3/libexec/nbd_client_manager.py (hunk 3 above)
Comment (stephenchengCloud) [reply python3/libexec/nbd_client_manager.py:48]: You're right. Python3.6 supports f-string.
I tried running the below f-string code directly on XS8, and it works:

```
[root@lcy2-dt49 ~]# python3 -V
Python 3.6.8
[root@lcy2-dt49 ~]# python3
Python 3.6.8 (default, Nov 16 2020, 16:55:22) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on Linux
Type "help", "copyright", "credits" or "license" for more information.
>>> nbd_device = "nbd2"
>>> f"NBD device '{nbd_device}' does not exist"
"NBD device 'nbd2' does not exist"
```
Now I have the same confusion as you, why the build error? I'm not the only one who saw this error. @acefei 
Suspect that during the build process, the environment doesn't support f-string. Need to investigate further.
```
Bytecompiling .py files below /builddir/build/BUILDROOT/xapi-24.11.0.21.gcb135e00f-1.1.g2e65c85.xs8.x86_64/usr/lib/python3.6 using /usr/bin/python3.6
Compiling /builddir/build/BUILDROOT/xapi-24.11.0.21.gcb135e00f-1.1.g2e65c85.xs8.x86_64/opt/xensource/libexec/nbd_client_manager.py ...
File "/opt/xensource/libexec/nbd_client_manager.py", line 48
f"NBD device '{nbd_device}' does not exist"
^
SyntaxError: invalid syntax
```

Comment (acefei) [reply python3/libexec/nbd_client_manager.py:48]: I encounted the similiar issue, but it failed as `py_byte_compile` running (https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#_source_files_and_bytecode_cache) both py2 compile and py3 compile against .py files in the current xapi.spec. in this case, f-string won't be accepted by py2 compiling.
```
+ /usr/lib/rpm/brp-python-bytecompile /usr/bin/python2 1
Bytecompiling .py files below /builddir/build/BUILDROOT/xapi-24.11.0.92.g86c1e653a-1.4.g7cc3472.xs8.x86_64/usr/lib/python2.7 using /usr/bin/python2.7
Bytecompiling .py files below /builddir/build/BUILDROOT/xapi-24.11.0.92.g86c1e653a-1.4.g7cc3472.xs8.x86_64/usr/lib/python3.6 using /usr/bin/python3.6
```

For @stephenchengCloud 's case, it is weird that I can't reproduce it on my mock env, and I tried it as follows
```
<mock-chroot> sh-4.2$ cat > 1.py
a='123'
print(f"{a}")
<mock-chroot> sh-4.2$ python3 ./1.py
123
<mock-chroot> sh-4.2$ ls
1.py  build
<mock-chroot> sh-4.2$ python3 -m py_compile 1.py
<mock-chroot> sh-4.2$ ls
1.py  __pycache__  build
<mock-chroot> sh-4.2$ cd __pycache__/
<mock-chroot> sh-4.2$ ls
1.cpython-36.pyc
<mock-chroot> sh-4.2$ ls
1.cpython-36.pyc
```

Comment (stephenchengCloud) [reply python3/libexec/nbd_client_manager.py:48]: I think my issue is the same as yours. I managed to reproduce the issue and updated the error log on my previous comment and it's the same as yours.

I will change the PR commit message to include this info.

**Comment by liulinC** [comment]
Yes, as @acefei  mentioned, the `f-string` issue is caused during rpm build as we has py2 lib support, and py2 byte compile. 
the py2 compile does not accept it.

It is just a string construction way, almost the same as `{}.format` (only a tiny performance diff, not a big deal).

File: python3/libexec/nbd_client_manager.py (hunk 2 above)
Comment (minglumlu) [reply python3/libexec/nbd_client_manager.py:62]: The `with` will be applied to `FileLock`.

File: python3/libexec/nbd_client_manager.py [hunk 4]
```diff
@@ -124,74 +134,101 @@ def _find_unused_nbd_device():
...
     """
     Return the full path for the persistent file containing
     the connection details. This is based on the device
     name, so /dev/nbd0 -> /var/run/nonpersistent/nbd/0
     """
-    number = re.search('/dev/nbd([0-9]+)', device).group(1)
-    return PERSISTENT_INFO_DIR + '/' + number
+    matched = re.search("/dev/nbd([0-9]+)", device)
+    if not matched:
+        raise NotGetNbdNumber("Can not get the nbd number")
```
Comment (minglumlu) [inline python3/libexec/nbd_client_manager.py:171]: `InvalidNbdDevName`?

File: python3/libexec/nbd_client_manager.py [hunk 5]
```diff
@@ -1,52 +1,66 @@
...
-LOCK_FILE = '/var/run/nonpersistent/nbd_client_manager'
+LOCK_FILE = "/var/run/nonpersistent/nbd_client_manager"
 
 # Don't wait more than 10 minutes for the NBD device
 MAX_DEVICE_WAIT_MINUTES = 10
 
 
+class NotGetNbdNumber(Exception):
+    """
+    The NBD device should be in this format: nbd{0-100}
```
Comment (minglumlu) [inline python3/libexec/nbd_client_manager.py:30]: The following line suggests there may be 1000 :(
`for device_no in range(0, 1000):`

File: python3/libexec/nbd_client_manager.py (hunk 4 above)
Comment (stephenchengCloud) [reply python3/libexec/nbd_client_manager.py]: Code updated

File: python3/libexec/nbd_client_manager.py [hunk 6]
```diff
@@ -124,74 +134,101 @@ def _find_unused_nbd_device():
         nbd_device = "/dev/nbd{}".format(device_no)
         if not _is_nbd_device_connected(nbd_device=nbd_device):
             return nbd_device
-
+    # Actually `_is_nbd_device_connected` will raise an exception
+    # if no unused device
+    # Add this return for pylint check
+    return None
```
Comment (minglumlu) [inline python3/libexec/nbd_client_manager.py:140]: Although it is impossible, but this return value would get the caller fail in error as the type of return doesn't match.

Comment (stephenchengCloud) [reply python3/libexec/nbd_client_manager.py:140]: It's impossible to have 1000 NBD devices and all of them are connected.
But you are right. Change this `return` to `raise exception`.

**Comment by minglumlu** [comment]
The running logs can be found in user.log of job 3944595 (Debian passed)
```
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['modprobe', 'nbd', 'nbds_max=24']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-unix', '/run/blktap-control/nbd/3475f067-8bc6-4f78-d475-98121c74fe1d/556756a1-56e7-4934-947c-ab3a085f9a1b', '/dev/nbd0', '-timeout', '60', '-name', '556756a1-56e7-4934-947c-ab3a085f9a1b']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-disconnect', '/dev/nbd0']
Mar  5 07:24:23 xrtmia-13-01 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
```

**Comment by stephenchengCloud** [comment]
Re-run XenRT Debian test based on the latest commit:
3955265 passed

user.log:
```
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['modprobe', 'nbd', 'nbds_max=24']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-unix', '/run/blktap-control/nbd/3edcce74-c574-8f7c-7a69-f86528890596/720ff3f4-0925-4d4d-a774-70596dd11d3e', '/dev/nbd0', '-timeout', '60', '-name', '720ff3f4-0925-4d4d-a774-70596dd11d3e']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-disconnect', '/dev/nbd0']
Mar 15 07:02:48 xrtmia-12-05 nbd_client_manager: [DEBUG] Running cmd ['nbd-client', '-check', '/dev/nbd0']
```

**Review by minglumlu — APPROVED** [review]

**Pull request merged**